chronos/bipbuffer.nim

```diff
@@ -0,0 +1,167 @@
+  let (res, _) = bp.calcReserve()
+  res
+
+func usedSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of used space in buffer `bp`.
+  len(bp.b) + len(bp.a)
+
+proc reserve*(bp: var BipBuffer, size: Natural): Result[Natural, cstring] =
+  ## Reserve `size` bytes in buffer. Returns number of bytes reserved.
+  doAssert(size <= len(bp.data))
+  let (availableSpace, reserveStart) = bp.calcReserve()
+  if availableSpace == 0:
+    return err("Not enough space available")
+  let reserveLength = min(availableSpace, size)
+  bp.r = BipPos.init(reserveStart, Natural(reserveStart + reserveLength))
+  ok(availableSpace)
+
+proc reserve*(bp: var BipBuffer): Natural =
+  ## Reserve all available free space in buffer. Returns number of bytes
+  ## reserved.
+  ##
+  ## Note, this procedure will raise Defect if there is no free space available.
+  let (availableSpace, reserveStart) = bp.calcReserve()
+  doAssert(availableSpace > 0, "Not enough space available")
+  bp.r = BipPos.init(reserveStart, Natural(reserveStart + availableSpace))
+  availableSpace
+
+proc reserve*(bp: var BipBuffer,
+              pt, st: typedesc): tuple[data: pt, size: st] =
+  ## Reserve all available free space in buffer. Returns current reserved range
+  ## as pointer of type `pt` and size of type `st`.
+  ##
+  ## Note, this procedure will raise Defect if there is no free space available.
+  let (availableSpace, reserveStart) = bp.calcReserve()
+  doAssert(availableSpace > 0, "Not enough space available")
+  bp.r = BipPos.init(reserveStart, Natural(reserveStart + availableSpace))
+  (cast[pt](addr bp.data[bp.r.start]), cast[st](len(bp.r)))
+
+func getReserve*(bp: var BipBuffer,
+                 pt: typedesc, st: typedesc): tuple[data: pointer, size: cint] =
+  ## Returns current reserved range as pointer + size tuple.
+  (cast[pt](addr bp.data[bp.r.start]), st(len(bp.r)))
+
+proc commit*(bp: var BipBuffer, size: Natural) =
+  ## Updates structure's pointers when new data inserted into buffer.
+  doAssert(len(bp.r) >= size,
+    "Committed size could not be larger than the previously reserved one")
+  if size == 0:
+    bp.r.zero()
+    return
+
+  let toCommit = min(size, len(bp.r))
+  if len(bp.a) == 0 and len(bp.b) == 0:
+    bp.a.start = bp.r.start
+    bp.a.finish = bp.r.start + toCommit
+  elif bp.r.start == bp.a.finish:
+    bp.a.finish += toCommit
+  else:
+    bp.b.finish += toCommit
+  bp.r.zero()
+
+proc decommit*(bp: var BipBuffer, size: Natural): Natural =
+  ## The procedure removes/frees `size` bytes from the buffer.
+  ## Returns number of bytes actually removed.
+  var currentSize = size
+  if currentSize >= len(bp.a):
+    currentSize -= len(bp.a)
+    bp.a = bp.b
+    bp.b.zero()
+    if currentSize >= len(bp.a):
+      currentSize -= len(bp.a)
+      bp.a.zero()
+      size - currentSize
+    else:
+      bp.a.start += currentSize
+      size
+  else:
+    bp.a.start += currentSize
+    size
+
+iterator bytes*(bp: BipBuffer): byte =
```

Comment:
```suggestion
iterator items*(bp: BipBuffer): byte =
```

allows it to be used in for loop easily

Reply:
Fixed in https://github.com/status-im/nim-chronos/pull/485/commits/e596388d5c2fa222433da97e96d72040473b2bcf